File: week14/15/2_Tue_SwapNodes.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return
        if head.next:
            head.val, head.next.val = head.next.val, head.val
            self.swapPairs(head.next.next)

        return head

File: week14/15/2_Tue_SwapNodes.cpp

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
 
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if (!head) return head;
        if (head->next) {
            int temp = head->val;
            head->val = head->next->val;
            head->next->val = temp;
            swapPairs(head->next->next);
        }
        return head;
    }
};